### PR TITLE
Pass dynamic workflow closure in node execution events

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Pass dynamic workflow closure in node execution events and also serialize as part of NodeExecutions in flyteadmin.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/831

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/event/event.proto [hunk 1]
```diff
@@ -93,6 +95,14 @@ message TaskNodeMetadata {
     core.CatalogMetadata catalog_key = 2;
 }
 
+// For dynamic workflow nodes we send information about the dynamic workflow definition that gets generated.
+message DynamicWorkflowNodeMetadata {
+    // id represents the unique identifier of the workflow.
+    core.Identifier id     = 1;
+
+    // Represents the compiled representation of the embedded dynamic workflow.
+    core.CompiledWorkflowClosure compiled_workflow = 2;
```
Comment (katrogan) [inline protos/flyteidl/event/event.proto:104]: not sure if we want this, might just save and return the workflow id only

File: protos/flyteidl/admin/node_execution.proto [hunk 2]
```diff
@@ -152,6 +154,15 @@ message TaskNodeMetadata {
     core.CatalogMetadata catalog_key = 2;
 }
 
+// For dynamic workflow nodes we send information about the dynamic workflow definition that gets generated.
+message DynamicWorkflowNodeMetadata {
+    // id represents the unique identifier of the workflow.
+    core.Identifier id     = 1;
+
+    // Represents the compiled representation of the embedded dynamic workflow.
+    core.CompiledWorkflowClosure compiled_workflow = 2;
```
Comment (katrogan) [inline protos/flyteidl/admin/node_execution.proto:163]: not sure if we want this, might just save and return the workflow id only

**Review by EngHabu — DISMISSED** [review]

File: protos/flyteidl/event/event.proto (hunk 1 above)
Comment (kumare3) [inline protos/flyteidl/event/event.proto:104]: is this what a regular workflow is also represented as?

Comment (katrogan) [reply protos/flyteidl/event/event.proto:104]: yes. https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/workflow.proto#L60

**Review by kumare3 — DISMISSED** [review]

File: protos/flyteidl/admin/node_execution.proto [hunk 3]
```diff
@@ -150,6 +151,18 @@ message TaskNodeMetadata {
     core.CatalogCacheStatus cache_status = 1;
     // This structure carries the catalog artifact information
     core.CatalogMetadata catalog_key = 2;
+
+    // In the case this task launched a dynamic workflow we capture its structure here.
+    DynamicWorkflowNodeMetadata dynamic_workflow_node_metadata = 16;
```
Comment (EngHabu) [inline protos/flyteidl/admin/node_execution.proto:156]: nit
```suggestion
    DynamicWorkflowNodeMetadata dynamic_workflow = 16;
```

File: protos/flyteidl/event/event.proto [hunk 4]
```diff
@@ -91,8 +92,19 @@ message TaskNodeMetadata {
     core.CatalogCacheStatus cache_status = 1;
     // This structure carries the catalog artifact information
     core.CatalogMetadata catalog_key = 2;
+
+    // In the case this task launched a dynamic workflow we capture its structure here.
+    DynamicWorkflowNodeMetadata dynamic_workflow_node_metadata = 16;
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:97]: nit:
```suggestion
    DynamicWorkflowNodeMetadata dynamic_workflow = 16;
```

**Review by EngHabu — APPROVED** [review]

File: protos/flyteidl/event/event.proto (hunk 4 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto]: @EngHabu done (didn't use your suggestion b/c DCO :( )

**Review by EngHabu — APPROVED** [review]

**Pull request merged**